### PR TITLE
NIFI-9768: Fix indeterminate map ordering in testSchemaToAttribute

### PR DESCRIPTION
**Pull request by MarcyGO**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

The test `org.apache.nifi.processors.standard.TestAttributesToCSV#testSchemaToAttribute` can fail based on the order of iteration in `HashMap`. I found the issue using [NonDex](https://github.com/TestingResearchIllinois/NonDex):
```
mvn install -DskipTests -pl nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors -am
mvn -pl nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors edu.illinois:nondex-maven-plugin:1.1.2:nondex -Dtest=org.apache.nifi.processors.standard.TestAttributesToCSV#testSchemaToContent
```
In the same test class, `testSchemaToContent`, `testSchemaWithCoreAttribuesToAttribute`, `testSchemaWithCoreAttribuesToContent` can also fail due to the same reason. 

The original proposed fix changes several `HashMap` to `LinkedHashMap`. But this change would cause  other test to fail: `org.apache.nifi.remote.protocol.http.TestHttpFlowFileServerProtocol.testTransferTwoFiles`. 

The current solution changes the assertion the endure the indeterminate behavior of HashMap. Instead of comparing strings, the commit converts the strings into lists using the existed function `getStrings`. Then, I wrote a helper function `zipToMap` that converts two lists into a map. It makes sure items in `CSVData` and `CSVSchema` are matched, but also ignores the order in the list. 

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [ ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by pvillard31** [comment]
Please file a JIRA and update your Pull Request title with the JIRA number. Thanks!

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/AttributesToCSV.java [hunk 1]
```diff
@@ -230,6 +230,8 @@ protected void init(final ProcessorInitializationContext context) {
         //now glue on the core attributes if the user wants them.
         if(includeCoreAttributes) {
             for (String coreAttribute : coreAttributes) {
+                System.out.println("****************** " + coreAttributes.getClass());
+                System.out.println(coreAttribute);
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/AttributesToCSV.java:234]: `System.out.println()` should not be used.
```suggestion
```

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/AttributesToCSV.java [hunk 2]
```diff
@@ -288,7 +290,9 @@ public void onTrigger(ProcessContext context, ProcessSession session) throws Pro
         }
 
         final Set<String> attributeList = attributeListStringToSet(context.getProperty(ATTRIBUTES_LIST).evaluateAttributeExpressions(original).getValue());
+        System.out.println("****************** attributeList " + attributeList.getClass());
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/AttributesToCSV.java:293]: `System.out.println()` should not be used.
```suggestion
```

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/AttributesToCSV.java [hunk 3]
```diff
@@ -288,7 +290,9 @@ public void onTrigger(ProcessContext context, ProcessSession session) throws Pro
...
         final Map<String, String> atrList = buildAttributesMapForFlowFile(original, attributeList, pattern);
+        System.out.println("****************** atrList " + atrList.getClass());
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/AttributesToCSV.java:295]:
```suggestion
```

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for the contribution @MarcyGO! The change from HashMap to LinkedHashMap in the test classes looks good, but `System.out.println()` calls should be removed.  Some tests leverage SLF4J Loggers, but in general, console output should not be necessary.

**Comment by MarcyGO** [comment]
Sorry, I accidentally push other commits to this branch.

**Comment by exceptionfactory** [comment]
Thanks for removing those lines @MarcyGO, I have enabled the automated build to run in order to verify the changes.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
@MarcyGO I committed a change to remove an empty file named `NOW`, but it appears that all automated checks are failing as a result of these changes.

The particular failing test is as follows:

```
Error:  org.apache.nifi.remote.protocol.http.TestHttpFlowFileServerProtocol.testTransferTwoFiles  Time elapsed: 0.661 s  <<< ERROR!
java.io.IOException: StandardHttpFlowFileServerProtocol[CommsID=testTransferTwoFiles] Sent data to peer Peer[url=http://peer-host:8080/] but calculated CRC32 Checksum as 2295235398 while peer calculated CRC32 Checksum as 3058746557; canceling transaction and rolling back session
	at org.apache.nifi.remote.protocol.http.TestHttpFlowFileServerProtocol.testTransferTwoFiles(TestHttpFlowFileServerProtocol.java:411)
```

Please run a full build and attempt to correct the issue, otherwise this pull request will not be able to go forward.

In general, it does not seem like these changes should result in new test failures, but the `CRC32 Checksum` error seems to indicate that this change alters the behavior of another test.

**Comment by MarcyGO** [comment]
Thank you for the feedback. I am working to fix the issue.

**Comment by exceptionfactory** [comment]
> Thank you for the feedback. I am working to fix the issue. Do you think, instead of the present adjustment, I should update the assertions in the test to withstand HashMap's unpredictable behavior? I implement the other solution in another branch, and all automated checks passes for that one.

Thanks for the reply @MarcyGO. If the failing test relies on incorrect assumptions about the behavior of HashMap, then it sounds like that particular test should be changed.  If you have a solution, please commit the change and we can evaluate it together with the current set of changes.

**Comment by MarcyGO** [comment]
Because you told me to change the test code, I've prepared this tentative PR (for `testSchemaWithCoreAttribuesToContent`): https://github.com/MarcyGO/nifi/pull/4  Please let me know if you like that change, and I can then open a real PR here on apache/nifi repo.

**Comment by MarcyGO** [comment]
I commit the changes which change assertions in four test functions ```testSchemaToAttribute```, ```testSchemaToContent```, ```testSchemaWithCoreAttribuesToAttribute```, and ```testSchemaWithCoreAttribuesToContent``` to endure the indeterminate behaviors of  ```HashMap```.

**Comment by MarcyGO** [comment]
I've edited the PR description.

**Comment by exceptionfactory** [comment]
Sorry for the delay in responding to the latest updates @MarcyGO.

Reviewing the most recent changes, I may not have communicated clearly in previous comments.

I understand that the current change set adjusts the behavior of `TestAttributesToCSV`, but it is does not address the problem described in the original version of the PR.

The this particular unit test class should remain unchanged, and it should continue to use the `assertAttributeEquals` method from `MockFlowFile`. If there is a problem with `assertAttributeEquals()`, then changes should be made to that method, and all impacted references should be updated. I realize that may involve changing other tests in addition to `TestAttributesToCSV`, but if that is necessary to correct the underlying behavior, that is a much better solution than changing individual tests to avoid using `assertAttributeEquals()`.

Although the `NonDex` tool reports an issue, we have not observed indeterminate behavior in normal test execution. There may still be opportunity for improvement, but it may be worth closing this particular PR and addressing the more fundamental issue.

Thanks again for your efforts!

**Comment by MarcyGO** [comment]
I can change `assertAttributeEquals()` so it ignore attribute order, or I can create another method `assertAttributeEqualsUnordered()` in `MockFlowFile` and use it to compare.

**Comment by exceptionfactory** [comment]
> I can change `assertAttributeEquals()` so it ignore attribute order, or I can create another method `assertAttributeEqualsUnordered()` in `MockFlowFile` and use it to compare.

The problem appears to be with the attribute value, so it looks like the `assertAttributeEquals()` method should remain unchanged.

Taking a closer look at `TestAttributesToCSV`, it looks like swapping out `HashMap` references with `LinkedHashMap` would address the non-deterministic issue.

**Comment by MarcyGO** [comment]
So I should keep the original version of PR (which change several `HashMap` to `LinkedHashMap`) and fix the problem that other two tests (`org.apache.nifi.remote.protocol.http.TestHttpFlowFileServerProtocol.testTransferTwoFiles` and `org.apache.nifi.remote.protocol.http.TestHttpFlowFileServerProtocol.testTransferOneFiles` ) would fail after the change.

**Comment by exceptionfactory** [comment]
> So I should keep the original version of PR (which change several `HashMap` to `LinkedHashMap`) and fix the problem that other two tests (`org.apache.nifi.remote.protocol.http.TestHttpFlowFileServerProtocol.testTransferTwoFiles` and `org.apache.nifi.remote.protocol.http.TestHttpFlowFileServerProtocol.testTransferOneFiles` ) would fail after the change.

Yes, going back to the original PR changes and fixing those other tests sounds like the best approach. Thanks for working through these details!

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
@MarcyGO, given the current state of the PR and the number of changes, it would be helpful to start with a new pull request, having a single commit, that attempts to resolve the issue apparently flagged with `assertAttributeEquals`. Please feel free to follow up with any question on the associated Jira issue and we can work through a new pull request when it is ready.

**Pull request closed without merging**